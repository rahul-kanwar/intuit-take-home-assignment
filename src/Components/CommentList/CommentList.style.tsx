import styled from "styled-components";

export const List = styled.div`
  border-top: 1px solid #ccc;
  padding-top: 20px;
`;