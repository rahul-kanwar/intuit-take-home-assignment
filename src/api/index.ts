import axios from 'axios';

export const apiClient = axios.create({
  baseURL: 'http://localhost:3001',
});

export interface Comment {
  id: number;
  name: string;
  created: string;
  message: string;
}

export const getComments = async (): Promise<Comment[]> => {
  const response = await apiClient.get('/getComments');
  return response.data;
};

export const createComment = async (comment: Omit<Comment, 'id'>): Promise<Comment> => {
  const response = await apiClient.post('/createComment', comment);
  return response.data;
};

export const deleteComments = async (): Promise<void> => {
  await apiClient.delete('/deleteComments');
};
