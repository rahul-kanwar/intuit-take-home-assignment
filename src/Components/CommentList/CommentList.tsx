import React, { useState, useEffect } from 'react';
import { getComments, Comment } from '../../api';
import CommentItem from '../CommentItem/CommentItem';
import { List } from './CommentList.style';

interface CommentListProps {
  newComment: Comment | null;
}

const CommentList: React.FC<CommentListProps> = ({ newComment }) => {
  const [comments, setComments] = useState<Comment[]>([]);

  useEffect(() => {
    const fetchComments = async () => {
      try {
        const data = await getComments();
        setComments(data);
      } catch (error) {
        console.error('Failed to fetch comments', error);
      }
    };
    fetchComments();
  }, []);

  useEffect(() => {
    if (newComment) {
      setComments((prevComments) => [...prevComments, newComment]);
    }
  }, [newComment]);

  return (
    <List>
      {comments.sort((a: Comment,b: Comment)=> b.id-a.id).map((comment) => (
        <CommentItem key={comment.id} comment={comment} />
      ))}
    </List>
  );
};

export default CommentList;
