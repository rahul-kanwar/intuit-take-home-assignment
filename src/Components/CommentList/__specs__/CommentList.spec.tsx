import React from 'react';
import { render, screen } from '@testing-library/react';
import 'jest-styled-components';
import CommentList from '../CommentList';
import { Comment } from '../../../api';


// Mock comments data
const mockComments: Comment[] = [
  {
    id: 1,
    name: 'John Doe',
    message: 'This is a test comment',
    created: '2024-06-26 10:35:42',
  },
  {
    id: 2,
    name: 'Jane Doe',
    message: 'Another test comment',
    created: '2024-06-26 11:35:42',
  },
];

describe('CommentList Component', () => {
  it('renders correctly without new comment', () => {
    const { asFragment } = render(<CommentList newComment={null} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with new comment', () => {
    const newComment: Comment = {
      id: 3,
      name: 'New User',
      message: 'This is a new comment',
      created: '2024-06-26 12:35:42',
    };
    const { asFragment } = render(<CommentList newComment={newComment} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays the list of comments', () => {
    render(<CommentList newComment={null} />);
    mockComments.forEach(comment => {
      expect(screen.getByText(comment.name)).toBeInTheDocument();
      expect(screen.getByText(comment.message)).toBeInTheDocument();
      expect(screen.getByText(/6\/26\/2024/)).toBeInTheDocument();
    });
  });
});
