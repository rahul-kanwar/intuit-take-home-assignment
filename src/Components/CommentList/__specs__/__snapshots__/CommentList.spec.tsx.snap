// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentList Component renders correctly with new comment 1`] = `
<DocumentFragment>
  .c1 {
  border-bottom: 1px solid #ccc;
  padding: 10px 0;
}

.c0 {
  border-top: 1px solid #ccc;
  padding-top: 20px;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <h4>
        New User
      </h4>
      <p>
        This is a new comment
      </p>
      <small>
        06/26/2024, 12:35 PM
      </small>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CommentList Component renders correctly without new comment 1`] = `
<DocumentFragment>
  .c0 {
  border-top: 1px solid #ccc;
  padding-top: 20px;
}

<div
    class="c0"
  />
</DocumentFragment>
`;
